src/Workspaces/Core/Portable/Workspace/Workspace_Registration.cs

```diff
@@ -56,7 +56,12 @@ protected void UnregisterText(SourceTextContainer textContainer)
                 throw new ArgumentNullException(nameof(textContainer));
             }
 
-            GetWorkspaceRegistration(textContainer).SetWorkspaceAndRaiseEvents(null);
+            var registration = GetWorkspaceRegistration(textContainer);
+
+            if (registration.Workspace == this)
```

Comment:
I don't think so. This isn't called directly, but rather indirectly when we are calling OnDocumentClosed once we see somebody else has it. Without threading through a bool that says "trust me, that's OK", I think this is safe.